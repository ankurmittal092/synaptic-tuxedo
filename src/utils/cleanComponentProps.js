import React from 'react';
import styled from 'styled-components';

const cleanComponentProps = (Component, blacklist = []) => {
  const removeBlackListProps = props => {
    const cleanedProps = blacklist.reduce((cProps, item) => {
      delete cProps[item];
      return cProps;
    }, Object.assign({}, props));

    return <Component {...cleanedProps} />;
  };

  return styled(removeBlackListProps);
};

export default cleanComponentProps;
