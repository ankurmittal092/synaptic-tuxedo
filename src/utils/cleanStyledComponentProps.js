import { omit } from 'clean-tag';
import React from 'react';
import PropTypes from 'prop-types';

// Workaround for the bug: https://github.com/jxnblk/styled-system/issues/241
const cleanComponent = (Component, additionalBlackList = []) => {
  const CleanedComponent = ({ blacklist = [], ...props }) => (
    <Component {...omit(props, [...blacklist, 'is', ...additionalBlackList])} />
  );
  CleanedComponent.propTypes = {
    blacklist: PropTypes.any
  };
  return CleanedComponent;
};

export default cleanComponent;
