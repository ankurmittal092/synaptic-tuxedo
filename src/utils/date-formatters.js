import moment from 'moment';
import _ from 'lodash';

function getFullDateGMTFormat(timestamp) {
  if (Number(timestamp) !== 'NaN') {
    return moment.utc(Number(timestamp)).format('DD MMM YYYY HH:mm:ss');
  }
  return timestamp;
}

function getFullDateLocalFormat(timestamp) {
  if (Number(timestamp) !== 'NaN') {
    return moment(Number(timestamp)).format('DD MMM YYYY HH:mm:ss ZZ');
  }
  return timestamp;
}

function getShortDateFormat(timestamp) {
  if (Number(timestamp) !== 'NaN') {
    return moment(Number(timestamp)).format('DD MMM YY');
  }
  return timestamp;
}

function getMonthFormat(timestamp) {
  if (Number(timestamp) !== 'NaN') {
    return moment(Number(timestamp)).format('MMM YY');
  }
  return timestamp;
}

function getYearFormat(timestamp) {
  if (Number(timestamp) !== 'NaN') {
    return moment(Number(timestamp)).format('YYYY');
  }
  return timestamp;
}

function getQuarterFormat(timestamp) {
  if (Number(timestamp) !== 'NaN') {
    return moment(Number(timestamp)).format('[Q]Q YY');
  }
  return timestamp;
}

function getIntervalFormat(prevDate, nextDate) {
  if (prevDate.year() !== nextDate.year()) {
    return `${prevDate.format('DD MMM YY')} - ${nextDate.format('DD MMM YY')}`;
  }
  if (prevDate.month() !== nextDate.month()) {
    return `${prevDate.format('DD MMM')} - ${nextDate.format('DD MMM YY')}`;
  }
  return `${prevDate.format('DD')} - ${nextDate.format('DD MMM YY')}`;
}

function getWeekFormat(timestamp) {
  if (Number(timestamp) === 'NaN') {
    return timestamp;
  }
  const date = moment(Number(timestamp));
  let nextDate = moment(Number(timestamp)).add(6, 'd');
  if (nextDate.isAfter(moment())) {
    nextDate = moment();
  }
  return getIntervalFormat(date, nextDate);
}

function getPastWeekFormat(timestamp) {
  if (Number(timestamp) === 'NaN') {
    return timestamp;
  }
  const date = moment(Number(timestamp));
  const prevDate = moment(Number(timestamp)).subtract(6, 'd');
  return getIntervalFormat(prevDate, date);
}

function getMinutesFormat(timestamp) {
  if (Number(timestamp) === 'NaN') {
    return timestamp;
  }
  return moment(Number(timestamp)).format('mm:ss');
}

function emptyFn(params) {
  return params;
}

const rollupDateFormatterMap = {
  quarter: getQuarterFormat,
  day: getShortDateFormat,
  week: getWeekFormat,
  year: getYearFormat,
  month: getMonthFormat
};

export function getDateFormatterForRollup(rollup) {
  return rollupDateFormatterMap[rollup] || emptyFn;
}

// hack- remove this
function getShortDateFormatHack(timestamp) {
  if (moment(timestamp).isValid()) {
    return moment(timestamp).format('DD MMM YY');
  }
  return timestamp;
}

function getWeekFormatHack(timestamp) {
  if (!moment(timestamp).isValid()) {
    return timestamp;
  }
  const date = moment(timestamp);
  let nextDate = moment(timestamp).add(6, 'd');
  if (nextDate.isAfter(moment())) {
    nextDate = moment();
  }
  return getIntervalFormat(date, nextDate);
}

function getMonthFormatHack(timestamp) {
  if (moment(timestamp).isValid()) {
    return moment(timestamp).format('MMM YY');
  }
  return timestamp;
}

const rollupDateFormatterMapHack = {
  day: getShortDateFormatHack,
  week: getWeekFormatHack,
  month: getMonthFormatHack
};

export function getDateFormatterForRollupHack(rollup) {
  return rollupDateFormatterMapHack[rollup] || emptyFn;
}

export default {
  getFullDateGMTFormat,
  getFullDateLocalFormat,
  getShortDateFormat,
  getMonthFormat,
  getYearFormat,
  getIntervalFormat,
  getQuarterFormat,
  getWeekFormat,
  getMinutesFormat,
  getPastWeekFormat
};
