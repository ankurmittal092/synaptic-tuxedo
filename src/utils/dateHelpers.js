import moment from 'moment';
import _ from 'lodash';

function getOptions(startDate, endDate, inc, unit, format, pluralForm) {
  let currDate = startDate;
  const options = [];

  if (!inc) {
    endDate = +moment(endDate).subtract(1, pluralForm);
  }

  while (currDate <= endDate) {
    options.push({
      timestamp: +moment(currDate).startOf(unit),
      value: moment(currDate).format(format)
    });
    currDate = +moment(currDate)
      .startOf(unit)
      .add(1, pluralForm);
  }

  options.sort((a, b) => b.timestamp - a.timestamp);

  return options;
}

export function getOptionsPeriodText(firstDate, period) {
  const momentDate = moment(firstDate, 'YYYY-MM-DD');
  if (!momentDate.isValid()) {
    return null;
  }

  if (period === 'quarter') {
    return momentDate.format('YYYY [Q]Q');
  } else if (period === 'month') {
    return momentDate.format('YYYY MMM');
  } else if (period === 'week') {
    return momentDate.format('YYYY-MM-DD');
  } else if (period === 'year') {
    return momentDate.format('YYYY');
  } else if (period === 'day') {
    return momentDate.format('YYYY-MM-DD');
  }
}

export function getOptionsPeriodTextFeed(firstDate, period) {
  const momentDate = moment(firstDate, 'YYYY-MM-DD');
  if (!momentDate.isValid()) {
    return null;
  }

  if (period === 'quarter') {
    return momentDate.format('[Q]Q YYYY');
  } else if (period === 'month') {
    return momentDate.format('MMM YYYY');
  } else if (period === 'week') {
    return momentDate.format('DD-MM-YYYY');
  } else if (period === 'year') {
    return momentDate.format('YYYY');
  } else if (period === 'day') {
    return momentDate.format('DD-MM-YYYY');
  }
}

export function getDayOptions(startDate, endDate, inc) {
  return getOptions(startDate, endDate, inc, 'day', 'YYYY-MM-DD', 'days');
}

export function getWeekOptions(startDate, endDate, inc) {
  return getOptions(startDate, endDate, inc, 'week', 'YYYY-MM-DD', 'weeks');
}

export function getYearOptions(startDate, endDate, inc) {
  return getOptions(startDate, endDate, inc, 'year', 'YYYY', 'years');
}

export function getQuarterOptions(startDate, endDate, inc) {
  startDate = +moment(startDate).startOf('quarter');
  return getOptions(startDate, endDate, inc, 'quarter', 'YYYY [Q]Q', 'Q');
}

export function getMonthOptions(startDate, endDate, inc) {
  return getOptions(startDate, endDate, inc, 'month', 'YYYY MMM', 'months');
}

export function getOptionsByRollup(startDate, endDate, rollup, optionType) {
  if (
    startDate ===
    moment(startDate, 'YYYY-MM-DD')
      .startOf(rollup)
      .format('YYYY-MM-DD')
  ) {
    startDate = +moment(startDate, 'YYYY-MM-DD');
  } else {
    startDate = +moment(startDate, 'YYYY-MM-DD')
      .add(1, rollup)
      .startOf(rollup);
  }

  if (
    endDate ===
    moment(endDate, 'YYYY-MM-DD')
      .endOf(rollup)
      .format('YYYY-MM-DD')
  ) {
    endDate = +moment(endDate, 'YYYY-MM-DD');
  } else {
    endDate = +moment(endDate, 'YYYY-MM-DD')
      .subtract(1, rollup)
      .startOf(rollup);
  }

  let options = [];
  if (rollup === 'day') {
    options = getDayOptions(startDate, endDate, true);
  } else if (rollup === 'week') {
    options = getWeekOptions(startDate, endDate, true);
  } else if (rollup === 'month') {
    options = getMonthOptions(startDate, endDate, true);
  } else if (rollup === 'quarter') {
    options = getQuarterOptions(startDate, endDate, true);
  } else if (rollup === 'year') {
    options = getYearOptions(startDate, endDate, true);
  }

  if (optionType === 'cohort') {
    const endDateTimestamp = +moment(endDate);
    const cohortEndDateTimestamp = +moment(options[0].timestamp)
      .add(1, rollup)
      .endOf(rollup);
    if (endDateTimestamp < cohortEndDateTimestamp) {
      options.shift();
    }
  }

  return _.map(options, opt => ({
    label: opt.value,
    value: moment(opt.timestamp).format('YYYY-MM-DD')
  }));
}

export function getRelativeOptionsByRollup(startDate, endDate, rollup) {
  const normalOptions = getOptionsByRollup(startDate, endDate, rollup);
  return _.reduce(
    normalOptions,
    (mem, opt) => {
      const n = moment(opt.value, 'YYYY-MM-DD').diff(moment(), rollup);
      if (n === 0) {
        return mem;
      }
      mem.push({
        label: n + 1 < 0 ? `n${n + 1}` : 'Latest (n)',
        value: String(n)
      });
      return mem;
    },
    []
  );
}

export function getDateRange(inp, rollup) {
  switch (rollup) {
    case 'year':
      return [
        moment(inp, 'YYYY')
          .startOf('year')
          .format('YYYY-MM-DD'),
        moment(inp, 'YYYY')
          .endOf('year')
          .format('YYYY-MM-DD')
      ];
    case 'quarter':
      return [
        moment(inp, 'YYYY [Q]Q')
          .startOf('quarter')
          .format('YYYY-MM-DD'),
        moment(inp, 'YYYY [Q]Q')
          .endOf('quarter')
          .format('YYYY-MM-DD')
      ];
    case 'week':
      return [
        moment(inp, 'YYYY-MM-DD')
          .startOf('week')
          .format('YYYY-MM-DD'),
        moment(inp, 'YYYY-MM-DD')
          .endOf('week')
          .format('YYYY-MM-DD')
      ];
    case 'day':
      return [
        moment(inp, 'YYYY-MM-DD')
          .startOf('day')
          .format('YYYY-MM-DD'),
        moment(inp, 'YYYY-MM-DD')
          .startOf('day')
          .format('YYYY-MM-DD')
      ];
    default:
      return [
        moment(inp, 'YYYY MMM')
          .startOf('month')
          .format('YYYY-MM-DD'),
        moment(inp, 'YYYY MMM')
          .endOf('month')
          .format('YYYY-MM-DD')
      ];
  }
}

/**
 * getDateArray - Function to get an array of dates from start to end
 *
 * DATE_FORMAT: YYYY-MM-DD
 * @param {string} start - Start date
 * @param {string} end - End date
 * @return {array} - Array of objects of dates from start date to end date
 */
export function getDateArray(start, end) {
  if (!start || !end) {
    return null;
  }

  if (moment(start) > moment(end)) {
    return null;
  }

  const dates = [
    {
      date: start,
      day: moment(start).day(),
      isWeekEnd: moment(start).day() === 0 || moment(start).day() === 6
    }
  ];

  let curDate = start;

  while (curDate) {
    curDate = moment(curDate)
      .add(1, 'day')
      .format('YYYY-MM-DD');
    if (curDate && moment(curDate) <= moment(end)) {
      dates.push({
        date: curDate,
        day: moment(curDate).day(),
        isWeekEnd: moment(curDate).day() === 0 || moment(curDate).day() === 6
      });
    } else {
      curDate = false;
    }
  }

  return dates;
}

/**
 * getMonthDates
 * @param {string} start - Start date
 * @return {array} - Objects Array of dates for month extracted from start date
 */
export function getMonthDates(start) {
  const monthStartWeekStart = moment(start)
    .startOf('week')
    .format('YYYY-MM-DD');
  const end = moment(start)
    .endOf('month')
    .format('YYYY-MM-DD');
  const monthEndWeekEnd = moment(end)
    .endOf('week')
    .format('YYYY-MM-DD');

  return getDateArray(monthStartWeekStart, monthEndWeekEnd);
}

export function getStartMomentObject(inp, rollup) {
  switch (rollup) {
    case 'year':
      return moment(inp, 'YYYY').startOf('year');
    case 'quarter':
      return moment(inp, 'YYYY [Q]Q').startOf('quarter');
    case 'week':
      return moment(inp, 'YYYY-MM-DD').startOf('week');
    case 'day':
      return moment(inp, 'YYYY-MM-DD').startOf('day');
    default:
      return moment(inp, 'YYYY MMM').startOf('month');
  }
}

export function getNOptions(orderBy, rollup, analysisPeriod) {
  const today = moment();
  const analysisPeriodM = getStartMomentObject(analysisPeriod, rollup);
  const diff = today.diff(analysisPeriodM, `${rollup}s`);
  return _.range(1, diff).map(el => ({
    value: el,
    label: String(el)
  }));
}

export function getNOptionsNew(rollup, analysisPeriod, endDate) {
  const today = moment(endDate, 'YYYY-MM-DD')
    .add(1, 'day')
    .startOf(rollup);
  const analysisPeriodM = moment(analysisPeriod).startOf(rollup);
  const diff = today.diff(analysisPeriodM, `${rollup}s`);
  return _.range(1, diff).map(el => ({
    value: String(el),
    label: String(el)
  }));
}

export function getLastBasePeriodDate(analysisPeriod, rollup) {
  return moment(analysisPeriod, 'YYYY-MM-DD')
    .subtract(1, rollup)
    .endOf(rollup)
    .format('YYYY-MM-DD');
}

export function getLastBasePeriodRelativeDate(analysisRelativeDate, rollup) {
  return moment()
    .subtract(-analysisRelativeDate + 1, rollup)
    .endOf(rollup)
    .format('YYYY-MM-DD');
}

export function getFirstAnalysisPeriodDate(analysisPeriod, rollup) {
  return moment(analysisPeriod, 'YYYY-MM-DD')
    .add(1, rollup)
    .startOf(rollup)
    .format('YYYY-MM-DD');
}

export function getDateDifferenceByRollup(analysisDate, baseDate, rollup) {
  const analDategetTime = new Date(analysisDate);
  const baseDategetTime = new Date(baseDate);
  const timeDiff = Math.abs(
    analDategetTime.getTime() - baseDategetTime.getTime()
  );
  const diffDays = Math.ceil(timeDiff / (1000 * 3600 * 24));

  const rollUpDivMap = {
    year: 365,
    quarter: 90,
    month: 30,
    week: 7,
    day: 1
  };

  const dividend = rollUpDivMap[rollup];
  return Math.round(diffDays / dividend);
}

export const yearCategoriesMap = {
  day: Array.from(Array(366)).map((el, index) => index + 1),
  week: Array.from(Array(52)).map((el, index) => index + 1),
  dayInMonth: Array.from(Array(31)).map((el, index) => index + 1),
  month: [
    'JAN',
    'FEB',
    'MAR',
    'APR',
    'MAY',
    'JUN',
    'JUL',
    'AUG',
    'SEP',
    'OCT',
    'NOV',
    'DEC'
  ],
  quarter: ['Q1', 'Q2', 'Q3', 'Q4']
};

const dateMap = [
  [moment().subtract(6, 'month'), moment()],
  [moment().subtract(12, 'month'), moment()],
  [moment().subtract(2, 'year'), moment()],
  [moment().subtract(3, 'year'), moment()],
  [moment('2000-01-01'), moment()]
];

export function getDateMap() {
  return {
    'Last 6 months': dateMap[0],
    'Last 12 months': dateMap[1],
    'Last 2 Years': dateMap[2],
    'Last 3 Years': dateMap[3],
    Max: dateMap[4]
  };
}

export function getDateRangeByIndex(index = 3) {
  return dateMap[index];
}
