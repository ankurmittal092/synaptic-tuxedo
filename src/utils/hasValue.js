export default function hasValue(val) {
  if (val === null || val === undefined || val === '') {
    return false;
  }
  return true;
}
