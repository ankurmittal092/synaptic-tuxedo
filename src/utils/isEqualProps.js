import _ from 'lodash';
export default function isEqualProps(nextProps, thisProps, keys) {
  if (keys) {
    return _.reduce(
      keys,
      (mem, key) => mem && _.isEqual(nextProps[key], thisProps[key]),
      true
    );
  } else {
    return _.isEqual(nextProps, thisProps);
  }
}

export function isEqualOmitFunction(nextProps, props) {
  return _.isEqual(
    _.omit(nextProps, _.functions(nextProps)),
    _.omit(props, _.functions(props))
  );
}
