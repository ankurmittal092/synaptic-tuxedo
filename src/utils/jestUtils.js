import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import Theme from 'theme';

export function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={Theme}>{component}</ThemeProvider>
  );
}
