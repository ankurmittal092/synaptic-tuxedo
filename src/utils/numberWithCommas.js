import _ from 'lodash';

export default function numberWithCommas(x) {
  let parts = [];
  if (x || _.isNumber(x)) {
    if (_.isNumber(x)) {
      parts = x.toString().split('.') || [''];
      parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    } else if (_.isString(x) && x.length > 0) {
      parts = x.split('.');
      parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    }

    return parts.join('.');
  }
  return null;
}
