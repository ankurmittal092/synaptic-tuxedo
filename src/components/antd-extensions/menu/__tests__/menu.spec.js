import React from 'react';
import { render } from 'enzyme';
import Menu from '../index.js';
import 'jest-styled-components';
import { renderWithTheme } from 'tuxedo/utils/jestUtils';

describe('Tuxedo_AntdExtensions - Menu - Rendering', () => {
  it('Default Menu', () => {
    const tree = renderWithTheme(
      <Menu>
        <Menu.Item key="mail">
          Navigation One
        </Menu.Item>
        <Menu.Item key="test">
          Navigation test
        </Menu.Item>
        <Menu.Item key="app" disabled>
          Navigation Two
        </Menu.Item>
      </Menu>
    );
    expect(tree).toMatchSnapshot();
  });
});
