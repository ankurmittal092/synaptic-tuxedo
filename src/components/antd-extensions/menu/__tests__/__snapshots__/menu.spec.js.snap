// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tuxedo_AntdExtensions - Menu - Rendering Default Menu 1`] = `
.c0.ant-menu {
  padding: px;
}

.c0.ant-dropdown-menu {
  padding: 8px;
}

.c1 {
  border-radius: 3px;
  font-weight: 500;
}

<ul
  className="ant-menu c0 ant-menu-light ant-menu-root ant-menu-vertical"
  role="menu"
  style={Object {}}
>
  <li
    aria-disabled={undefined}
    className="ant-menu-item c1"
    onBlur={undefined}
    onClick={[Function]}
    onContextMenu={undefined}
    onFocus={undefined}
    onMouseDown={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={undefined}
    role="menuitem"
    style={Object {}}
    title={undefined}
  >
    Navigation One
  </li>
  <li
    aria-disabled={undefined}
    className="ant-menu-item c1"
    onBlur={undefined}
    onClick={[Function]}
    onContextMenu={undefined}
    onFocus={undefined}
    onMouseDown={undefined}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onTouchStart={undefined}
    role="menuitem"
    style={Object {}}
    title={undefined}
  >
    Navigation test
  </li>
  <li
    aria-disabled={true}
    className="ant-menu-item c1 ant-menu-item-disabled"
    onBlur={undefined}
    onClick={null}
    onContextMenu={undefined}
    onFocus={undefined}
    onMouseDown={undefined}
    onMouseEnter={null}
    onMouseLeave={null}
    onTouchStart={undefined}
    role="menuitem"
    style={Object {}}
    title={undefined}
  >
    Navigation Two
  </li>
</ul>
`;
